src/custom/state/orders/updaters/APIOrdersUpdater.ts

```diff
@@ -0,0 +1,116 @@
+      return OrderStatus.EXPIRED
+    case 'pending':
+      return OrderStatus.PENDING
+    case 'fulfilled':
+      return OrderStatus.FULFILLED
+    default:
+      return undefined
+  }
+}
+
+function transformApiOrderToStoreOrder(
+  order: OrderMetaData,
+  chainId: ChainId,
+  allTokens: { [address: string]: Token }
+): Order | undefined {
+  const { uid: id, sellToken, buyToken, creationDate: creationTime, receiver } = order
+
+  // TODO: load tokens from network when not found on local state
+  const inputToken = getToken(sellToken, chainId, allTokens)
+  const outputToken = getToken(buyToken, chainId, allTokens)
+
+  const apiStatus = classifyOrder(order)
+  const status = classifyLocalStatus(apiStatus)
+
+  if (!status) {
+    console.warn(`APIOrdersUpdater::Order ${id} in unknown internal state: ${apiStatus}`)
+    return
+  }
+
+  if (inputToken && outputToken) {
+    const storeOrder: Order = {
+      ...order,
+      inputToken,
+      outputToken,
+      id,
+      creationTime,
+      summary: '',
+      status,
+      receiver,
+      apiAdditionalInfo: order,
+      isCancelling: apiStatus === 'pending' && order.invalidated, // already cancelled in the API, not yet in the UI
+    }
+    // The function to compute the summary needs the Order instance to exist already
+    // That's why it's not used before and an empty string is set instead
+    storeOrder.summary = computeOrderSummary({ orderFromStore: storeOrder, orderFromApi: order }) || ''
+
+    return storeOrder
+  } else {
+    console.warn(
+      `APIOrdersUpdater::Tokens not found for order ${id}: sellToken ${!inputToken ? sellToken : 'found'} - buyToken ${
+        !outputToken ? buyToken : 'found'
+      }`
+    )
+    return
+  }
+}
+
+export function APIOrdersUpdater(): null {
+  const { account, chainId } = useActiveWeb3React()
+  const allTokens = useAllTokens()
+  const tokenAreLoaded = useMemo(() => Object.keys(allTokens).length > 0, [allTokens])
+  const addOrUpdateOrdersBatch = useAddOrUpdateOrdersBatch()
+
+  useEffect(() => {
+    if (account && chainId && tokenAreLoaded) {
+      getOrders(chainId, account, 100)
+        .then((apiOrders) => {
+          console.log(`APIOrdersUpdater::Fetched ${apiOrders.length} orders for account ${account} on chain ${chainId}`)
+
+          // Transform API orders into internal order objects and filter out orders that are not in a known state
+          const orders = apiOrders.reduce<Order[]>((acc, order) => {
+            const storeOrder = transformApiOrderToStoreOrder(order, chainId, allTokens)
+            if (storeOrder) {
+              acc.push(storeOrder)
+            }
+            return acc
+          }, [])
+
+          // Add to redux state
+          addOrUpdateOrdersBatch({ orders, chainId })
+        })
+        .catch((e) => {
+          console.error(`APIOrdersUpdater::Failed to fetch orders`, e)
+        })
+    }
+  }, [account, addOrUpdateOrdersBatch, allTokens, chainId, tokenAreLoaded])
```

Comment:
I wouldn't say intended but it's a side effect.
Trying to avoid as much as possible to fetch tokens from the network.

Any suggestion to better handle this?

Reply:
Just saw the suggestion on general comment.

I'll give that a try, but not on this PR. I'll do it either on https://github.com/gnosis/cowswap/pull/1504 or as follow up to it.